api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,205 @@
+#pragma once
+
+#include <atomic>
+
+#include "opentelemetry/context/context_value.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+// The context class provides a context identifier. Is built as a linked list
+// of context nodes which each have access to the nodes before it, but not 
+// after. So each time a write is performed a new context is added to the list.
+// Within each context node there is another linked list which holds the key
+// and value data for each node.
+class Context
+{
+
+public:
+  
+  // A linked list to contain the keys and values of this context node 
+  class DataList {
+    private:
+      friend class Context;
+ 
+      DataList *next_;
+      
+      char key_[50];
+      
+      ContextValue value_;
+  };
+  
+  Context(){
+    next_ = NULL;
+    head_ = NULL;
+  }
+
+  // Creates a context object from a map of keys and identifiers, this will
+  // be the head of the context object linked list.
+  template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context(const T &keys_and_values)  
+  {
+    head_ = BuildList(keys_and_values);
+    next_ = NULL;
+  }
+  
+  // Creates a context object from a key and value, this will be the head
+  // of the context object linked list.
+  Context(nostd::string_view key, ContextValue value)  
+  {
+    head_ = BuildHead(key, value);
+    next_ = NULL;
+  }
+
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data.
+  template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context* SetValues(T &values) noexcept
+  {
+    Context *new_node = new Context(values, this);
+    return new_node;
+  }
+  
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data.
+  Context* SetValue(nostd::string_view key, ContextValue value) noexcept
+  {
+    Context *new_node = new Context(key,value, this);
+    return new_node;
+  }
+
+  // Returns the value associated with the passed in key.
+  common::AttributeValue GetValue(nostd::string_view key)
+  {
+    Context *context_trace = this;
+    
+    //Iterate through the context nodes
+    while(context_trace != NULL){
+      DataList *data_trace = context_trace->head_;
+      
+      //Iterate through the internal data nodes
+      while(data_trace != NULL){
```

Comment:
WDYT about making these for loops?

  for (DataList* data = context->head_; data != nullptr; data = data->next_)

The advantage being that the index variable's scope and behavior is more tightly bound to the loop.

Reply:
Changed to for loops.